### PR TITLE
Translator support for SPV_INTEL_fpga_argument_interfaces

### PR DESCRIPTION
**Pull request by broxigarchen**
Regarding PR for spec https://github.com/KhronosGroup/SPIRV-Registry/pull/177

added intel SPV_INTEL_argument_interfaces Translator change.

### CODE REVIEW

**Comment by broxigarchen** [comment]
Will add internal headers to get it compile. And and test files soon

**Review by MrSidims — COMMENTED** [review]
Thanks for your contribution!
Also thanks for applying clang-format to the existing code, but it's a bit complicates reviewing. Could it be done in a separate patch?
There are several things missing in the patch:
1. Please add the extension to https://github.com/KhronosGroup/SPIRV-LLVM-Translator/blob/main/include/LLVMSPIRVExtensions.inc
2. Please add decorations to the list of those, which require an extension here: https://github.com/KhronosGroup/SPIRV-LLVM-Translator/blob/main/lib/SPIRV/libSPIRV/SPIRVDecorate.h#L130

**Comment by broxigarchen** [comment]
> Thanks for your contribution! Also thanks for applying clang-format to the existing code, but it's a bit complicates reviewing. Could it be done in a separate patch? There are several things missing in the patch:
> 
> 1. Please add the extension to https://github.com/KhronosGroup/SPIRV-LLVM-Translator/blob/main/include/LLVMSPIRVExtensions.inc
> 2. Please add decorations to the list of those, which require an extension here: https://github.com/KhronosGroup/SPIRV-LLVM-Translator/blob/main/lib/SPIRV/libSPIRV/SPIRVDecorate.h#L130

Thanks for pointing this out. Done!

File: test/sycl-kernel-arg-annotation.ll [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+; CHECK: ![[LATENCY]] = !{i32 6179, i32 1}
+; CHECK: ![[MAXBURST]] = !{i32 6181, i32 3}
+; CHECK: ![[RWMODE]]  = !{i32 6180, i32 2}
+; CHECK: ![[REGMAP]]  = !{i32 6176, i32 1}
+; CHECK: ![[STABLE]]  = !{i32 6184, i32 1}
+; CHECK: ![[STRICT]]  = !{i32 19, i32 1}
+; CHECK: ![[WAITREQ]] = !{i32 6183, i32 5}
+
+
+; CHECK-SPIRV-DAG:  Name [[ID:[0-9]+]] "_arg_p"
```
[the remark this answers is not in the document]
Comment (broxigarchen) [reply test/sycl-kernel-arg-annotation.ll:54]: done

File: test/sycl-kernel-arg-annotation.ll [hunk 2]
```diff
@@ -0,0 +1,66 @@
+; RUN: llvm-as %s -o %t.bc
+; RUN: llvm-spirv %t.bc -spirv-ext=-all,+SPV_INTEL_fpga_buffer_location -o %t.spv
```
[the remark this answers is not in the document]
Comment (broxigarchen) [reply test/sycl-kernel-arg-annotation.ll]: done

File: test/sycl-kernel-arg-annotation.ll [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+!1594 = !{i32 6178, i32 64}
+!1595 = !{i32 6179, i32 1}
+!1596 = !{i32 6181, i32 3}
+!1597 = !{i32 6180, i32 2}
+!1598 = !{i32 6176, i32 1}
+!1599 = !{i32 6184, i32 1}
+!1600 = !{i32 19, i32 1}
+!1601 = !{i32 6183, i32 5}
+
+; CHECK-DAG: ![[PARMDECOR_CASE1]] = !{![[ARG:[0-9]+]]}
```
[the remark this answers is not in the document]
Comment (broxigarchen) [reply test/sycl-kernel-arg-annotation.ll]: done

File: test/sycl-kernel-arg-annotation.ll [hunk 4]
```diff
@@ -0,0 +1,66 @@
+; RUN: llvm-as %s -o %t.bc
```
Comment (MrSidims) [inline test/sycl-kernel-arg-annotation.ll:1]: Please move the test to test/extensions/INTEL/SPV_INTEL_fpga_argument_interfaces directory

File: test/sycl-kernel-arg-annotation.ll [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+; RUN: llvm-spirv %t.bc -spirv-ext=-all,+SPV_INTEL_fpga_argument_interfaces -o %t.spv
+; RUN: llvm-spirv %t.spv -to-text -o - | FileCheck %s --check-prefix=CHECK-SPIRV
+; RUN: llvm-spirv -r %t.spv -o %t.rev.bc
+; RUN: llvm-dis < %t.rev.bc | FileCheck %s --check-prefix=CHECK-LLVM
+
+target datalayout = "e-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024-n8:16:32:64"
+target triple = "spir64_fpga-unknown-unknown"
```
Comment (MrSidims) [inline test/sycl-kernel-arg-annotation.ll:8]:
```suggestion
target triple = "spir64-unknown-unknown"
```

File: test/sycl-kernel-arg-annotation.ll [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+; CHECK-SPIRV:  Decorate [[ID]] MMHostInterfaceAddressWidthINTEL 32
+; CHECK-SPIRV:  Decorate [[ID]] BufferLocationINTEL 10
+; CHECK-SPIRV:  Decorate [[ID]] ConduitKernelArgumentINTEL
+; CHECK-SPIRV:  Decorate [[ID]] MMHostInterfaceDataWidthINTEL 64
+; CHECK-SPIRV:  Decorate [[ID]] MMHostInterfaceLatencyINTEL 1
+; CHECK-SPIRV:  Decorate [[ID]] MMHostInterfaceMaxBurstINTEL 3
+; CHECK-SPIRV:  Decorate [[ID]] MMHostInterfaceReadWriteModeINTEL 2
+; CHECK-SPIRV:  Decorate [[ID]] RegisterMapKernelArgumentINTEL
+; CHECK-SPIRV:  Decorate [[ID]] StableKernelArgumentINTEL
+; CHECK-SPIRV:  Decorate [[ID]] Restrict
```
Comment (MrSidims) [inline test/sycl-kernel-arg-annotation.ll:65]: Do we need it?

File: test/sycl-kernel-arg-annotation.ll (hunk 4 above)
Comment (broxigarchen) [reply test/sycl-kernel-arg-annotation.ll]: Done

File: test/sycl-kernel-arg-annotation.ll (hunk 5 above)
Comment (broxigarchen) [reply test/sycl-kernel-arg-annotation.ll]: done

File: lib/SPIRV/libSPIRV/SPIRVDecorate.h [hunk 7]
```diff
@@ -178,6 +178,17 @@ class SPIRVDecorate : public SPIRVDecorateGeneric {
     case internal::DecorationInitModeINTEL:
     case internal::DecorationImplementInCSRINTEL:
       return ExtensionID::SPV_INTEL_global_variable_decorations;
+    case DecorationConduitKernelArgumentINTEL:
+    case DecorationRegisterMapKernelArgumentINTEL:
+    case DecorationStableKernelArgumentINTEL:
+    case DecorationRestrict:
```
[the remark this answers is not in the document]
Comment (broxigarchen) [reply lib/SPIRV/libSPIRV/SPIRVDecorate.h]: This is an error I made. Corrected

File: test/sycl-kernel-arg-annotation.ll (hunk 6 above)
Comment (broxigarchen) [reply test/sycl-kernel-arg-annotation.ll]: Yes we need it here. We are reusing the "Restrict" property

File: lib/SPIRV/SPIRVWriter.cpp [hunk 8]
```diff
@@ -2433,6 +2433,30 @@ static void transMetadataDecorations(Metadata *MD, SPIRVEntry *Target) {
       TWO_INT_DECORATION_CASE(FuseLoopsInFunctionINTEL, spv, SPIRVWord,
                               SPIRVWord);
       TWO_INT_DECORATION_CASE(MathOpDSPModeINTEL, spv, SPIRVWord, SPIRVWord);
+
+    case DecorationConduitKernelArgumentINTEL:
+    case DecorationRegisterMapKernelArgumentINTEL:
+    case DecorationStableKernelArgumentINTEL:
+    case DecorationRestrict: {
```
[the remark this answers is not in the document]
Comment (broxigarchen) [reply lib/SPIRV/SPIRVWriter.cpp]: This is a required change and here is reusing the "Restrict" property which is added before

**Review by MrSidims — APPROVED** [review]

File: lib/SPIRV/SPIRVWriter.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (MrSidims) [reply lib/SPIRV/SPIRVWriter.cpp]: Not sure if I understand, this extension doesn't reuse "restrict" from what I see from the spec.

But I can imagine, that whiteout the code you add here neither restrict nor buffer location works properly.

**Comment by broxigarchen** [comment]
Updated with git-clang-format

File: lib/SPIRV/SPIRVWriter.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (broxigarchen) [reply lib/SPIRV/SPIRVWriter.cpp:2440]: The spec added a few new decorations under SPV_INTEL_fpga_argument_interfaces. The "Restrict" and "Bufferlocation" are defined in other specs (SPIRV spec and SPV_INTEL_fpga_buffer_location extension ) so they aren't added as new properties. But they are used to decorate some of the kernel arguments here.

**Pull request merged**